Makefile

```diff
@@ -64,32 +78,72 @@ clean_develop: check_venv
 	- $(pip) uninstall -y toil
 	- rm -rf src/*.egg-info
 
-
-sdist: check_venv
+sdist: dist/$(sdist_name)
+dist/$(sdist_name): check_venv
+	@test -f dist/$(sdist_name) && mv dist/$(sdist_name) dist/$(sdist_name).old || true
 	$(python) setup.py sdist
+	@test -f dist/$(sdist_name).old \
+	    && ( cmp -s <(tar -xOzf dist/$(sdist_name)) <(tar -xOzf dist/$(sdist_name).old) \
+	         && mv dist/$(sdist_name).old dist/$(sdist_name) \
+	         && printf "$(green)No significant changes to sdist, reinstating backup.$(normal)" \
+	         || rm dist/$(sdist_name).old ) \
+	    || true
 clean_sdist:
 	- rm -rf dist
 
 
 test: check_venv check_build_reqs
 	$(python) run_tests.py test $(tests)
 
-integration-test: check_venv check_build_reqs sdist
-	TOIL_TEST_INTEGRATIVE=True $(python) run_tests.py integration-test $(tests)
+
+integration-test: check_venv check_build_reqs sdist push_docker
+	TOIL_TEST_INTEGRATIVE=True \
+	TOIL_APPLIANCE_SELF=$(docker_registry)/$(docker_base_name):$(docker_tag) \
+	    $(python) run_tests.py integration-test $(tests)
 
 
 pypi: check_venv check_clean_working_copy check_running_on_jenkins
 	set -x \
 	&& tag_build=`$(python) -c 'pass;\
-		from version import version as v;\
-		from pkg_resources import parse_version as pv;\
-		import os;\
-		print "--tag-build=.dev" + os.getenv("BUILD_NUMBER") if pv(v).is_prerelease else ""'` \
+	    from version import version as v;\
+	    from pkg_resources import parse_version as pv;\
+	    import os;\
+	    print "--tag-build=.dev" + os.getenv("BUILD_NUMBER") if pv(v).is_prerelease else ""'` \
 	&& $(python) setup.py egg_info $$tag_build sdist bdist_egg upload
 clean_pypi:
 	- rm -rf build/
 
 
+docker: check_docker_registry docker/worker/Dockerfile docker/leader/Dockerfile
+	@set -ex \
+	; cd docker \
+	; for role in leader worker; do \
+	    docker build --tag=$(docker_registry)/$(docker_base_name)-$${role}:$(docker_tag) \
+	                 -f $${role}/Dockerfile \
+	                 . \
+	; done
+	docker tag -f $(docker_registry)/$(docker_base_name)-leader:$(docker_tag) \
+	              $(docker_registry)/$(docker_base_name):$(docker_tag)
```

Comment:
Shouldn't this be `$(docker_registry)/$(docker_base_name)-worker:$(docker_tag)`?